cpp/src/arrow/engine/api.h

```diff
@@ -19,6 +19,4 @@
 
 #pragma once
 
-#include "arrow/engine/substrait/extension_set.h"
 #include "arrow/engine/substrait/extension_types.h"
-#include "arrow/engine/substrait/serde.h"
```

Comment:
Why remove `serde.h`?  Is it transitively pulling in something internal or was this change a preference thing?  This is probably the most significant public header so I'll need to figure out how we can get it in this list.

Reply:
It is transitively pulling some internals. As I said the solution here deserved to be improved :-)

Reply:
Ok, I'll make an attempt to clean up exec & engine soon but we should keep going with this PR as-is.  Thanks for noticing this.

Reply:
https://github.com/apache/arrow/pull/14131